scipy/special/xsf/tools.h

```diff
@@ -261,10 +258,160 @@ namespace detail {
      * See ContinuedFractionSeriesGenerator for details.
      */
     template <typename Generator, typename T = pair_value_t<generator_result_t<Generator>>>
-    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T>
-    continued_fraction_series(Generator &cf) {
+    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T> continued_fraction_series(Generator &cf) {
         return ContinuedFractionSeriesGenerator<Generator, T>(cf);
     }
 
+    /* Find initial bracket for a bracketing scalar root finder. A valid bracket is a pair of points a < b for
+     * which the signs of f(a) and f(b) differ. x_left and x_right give the initial bracket, and x_min and
+     * x_max define the search space. This is a private function intended specifically for the situation where
+     * the goal is to invert a CDF function F for a parametrized family of distributions with respect to one
+     * parameter, when the other parameters are known, and where F is monotonic with respect to the unknown parameter.
+     *
+     * Note that this takes a pointer to a function taking a tuple of args along with a scalar
+     * double argument. A tuple of args for specializing func is also passed as the final argument.
+     * It would be much cleaner to use std::function and capturing lambda's to specialize the
+     * function we are finding a root for, but this was found unworkable in CuPy using NVRTC.
+     * This should be revisited in the future in order to allow simplifying this code. */
+    template <typename... Args>
+    XSF_HOST_DEVICE inline std::tuple<double, double, double, double, int> bracket_root_for_cdf_inversion(
```

Comment:
Just a comment (not a request for change): since the monotonicity of (the sign of) the function is known, unless the initial bracket is a correct guess, one of the two initial evaluations would be wasted. (Of course we don’t know which one in advance.) If, instead of passing an initial bracket, an initial guess x0 and an initial step size is passed, then the second evaluation would only need to go in the “right” direction. On average I think this would save half evaluation.

Reply:
This is a good suggestion. I agree we should do this to save the half function evaluation.